lib/uri_parser.js

```diff
@@ -384,6 +402,74 @@ function parseQueryString(query, options) {
   return Object.keys(result).length ? result : null;
 }
 
+/**
+ * Checks a query string for invalid tls options according to the URI options spec.
+ *
+ * @param {string} queryString The query string to check
+ * @throws {MongoParseError}
+ */
+function checkTLSOptions(queryString) {
+  const queryStringKeys = Object.keys(queryString);
+  if (
+    queryStringKeys.indexOf('tlsInsecure') !== -1 &&
+    (queryStringKeys.indexOf('tlsAllowInvalidCertificates') !== -1 ||
+      queryStringKeys.indexOf('tlsAllowInvalidHostnames') !== -1)
+  ) {
+    throw new MongoParseError(
+      'The tlsInsecure option cannot be used with tlsAllowInvalidCertificates or tlsAllowInvalidHostnames.'
+    );
+  }
+
+  const queryStringHasTLS = queryStringKeys.indexOf('tls') !== -1;
```

Comment:
starting here it looks like this can all be factored into a method that takes in either `ssl` or `tls` and runs the checks:

```
function checkTlsValues(type) {
  ...
}

checkTlsValues('tls');
checkTlsValues('ssl');
```

Reply:
additional nit: it seems more like we're "asserting" here rather than just "checking"